File: app/src/main/java/linusfessler/alarmtiles/shared/TimeFormatter.java
package linusfessler.alarmtiles.shared;

import java.util.Locale;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class TimeFormatter {

    @Inject
    public TimeFormatter() {

    }

    public String format(final long millis, final TimeUnit precision) {
        long millisLeft = millis;

        final long hours = TimeUnit.MILLISECONDS.toHours(millisLeft);
        millisLeft -= TimeUnit.HOURS.toMillis(hours);

        if (precision == TimeUnit.HOURS) {
            return formatHours(hours);
        }

        final long minutes = TimeUnit.MILLISECONDS.toMinutes(millisLeft);
        millisLeft -= TimeUnit.MINUTES.toMillis(minutes);

        if (precision == TimeUnit.MINUTES) {
            return format(hours, minutes);
        }

        final long seconds = TimeUnit.MILLISECONDS.toSeconds(millisLeft);
        millisLeft -= TimeUnit.SECONDS.toMillis(seconds);

        if (precision == TimeUnit.SECONDS) {
            return format(hours, minutes, seconds);
        }

        if (precision == TimeUnit.MILLISECONDS) {
            return format(hours, minutes, seconds, millisLeft);
        }

        throw new IllegalArgumentException(String.format("Precision %s is not supported.", precision));
    }

    public String format(final long hours, final long minutes, final long seconds, final long millis) {
        if (hours == 0 && minutes == 0 && seconds == 0) {
            return formatMillis(millis);
        }

        if (hours == 0 && minutes == 0) {
            return formatSeconds(seconds) + " " + formatMillis(millis);
        }

        if (hours == 0) {
            return formatMinutes(minutes) + " " + formatSeconds(seconds) + " " + formatMillis(millis);
        }

        return formatHours(hours) + " " + formatMinutes(minutes) + " " + formatSeconds(seconds) + " " + formatMillis(millis);
    }

    public String format(final long hours, final long minutes, final long seconds) {
        if (hours == 0 && minutes == 0) {
            return formatSeconds(seconds);
        }

        if (hours == 0) {
            return formatMinutes(minutes) + " " + formatSeconds(seconds);
        }

        return formatHours(hours) + " " + formatMinutes(minutes) + " " + formatSeconds(seconds);
    }

    public String format(final long hours, final long minutes) {
        if (hours == 0) {
            return formatMinutes(minutes);
        }

        return formatHours(hours) + " " + formatMinutes(minutes);
    }

    public String formatHours(final long hours) {
        return hours + "h";
    }

    public String formatMinutes(final long minutes) {
        return minutes + "m";
    }

    public String formatSeconds(final long seconds) {
        return seconds + "s";
    }

    public String formatMillis(final long millis) {
        return String.format(Locale.getDefault(), "%02d", millis / 10);
    }
}

File: app/src/main/java/linusfessler/alarmtiles/shared/TimeFormatter.kt
package linusfessler.alarmtiles.shared

import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class TimeFormatter @Inject constructor() {
    fun format(millis: Long, precision: TimeUnit): String {
        var millisLeft = millis

        val hours = TimeUnit.MILLISECONDS.toHours(millisLeft)
        millisLeft -= TimeUnit.HOURS.toMillis(hours)

        if (precision == TimeUnit.HOURS) {
            return formatHours(hours)
        }

        val minutes = TimeUnit.MILLISECONDS.toMinutes(millisLeft)
        millisLeft -= TimeUnit.MINUTES.toMillis(minutes)

        if (precision == TimeUnit.MINUTES) {
            return format(hours, minutes)
        }

        val seconds = TimeUnit.MILLISECONDS.toSeconds(millisLeft)
        millisLeft -= TimeUnit.SECONDS.toMillis(seconds)

        if (precision == TimeUnit.SECONDS) {
            return format(hours, minutes, seconds)
        }

        if (precision == TimeUnit.MILLISECONDS) {
            return format(hours, minutes, seconds, millisLeft)
        }

        throw IllegalArgumentException("Precision $precision is not supported.")
    }

    fun format(hours: Long, minutes: Long, seconds: Long, millis: Long): String {
        if (hours == 0L && minutes == 0L && seconds == 0L) {
            return formatMillis(millis)
        }

        if (hours == 0L && minutes == 0L) {
            return formatSeconds(seconds) + " " + formatMillis(millis)
        }

        return if (hours == 0L) {
            formatMinutes(minutes) + " " + formatSeconds(seconds) + " " + formatMillis(millis)
        } else {
            formatHours(hours) + " " + formatMinutes(minutes) + " " + formatSeconds(seconds) + " " + formatMillis(millis)
        }
    }

    fun format(hours: Long, minutes: Long, seconds: Long): String {
        if (hours == 0L && minutes == 0L) {
            return formatSeconds(seconds)
        }
        
        return if (hours == 0L) {
            formatMinutes(minutes) + " " + formatSeconds(seconds)
        } else {
            formatHours(hours) + " " + formatMinutes(minutes) + " " + formatSeconds(seconds)
        }
    }

    fun format(hours: Long, minutes: Long): String {
        return if (hours == 0L) {
            formatMinutes(minutes)
        } else {
            formatHours(hours) + " " + formatMinutes(minutes)
        }
    }

    fun formatHours(hours: Long): String {
        return hours.toString() + "h"
    }

    fun formatMinutes(minutes: Long): String {
        return minutes.toString() + "m"
    }

    fun formatSeconds(seconds: Long): String {
        return seconds.toString() + "s"
    }

    fun formatMillis(millis: Long): String {
        return String.format("%02d", millis / 10)
    }
}